lighthouse-core/gather/computed/dependency-graph/estimator/estimator.js

```diff
@@ -0,0 +1,291 @@
+    timingData.estimatedTimeElapsed = estimate;
+    return estimate;
+  }
+
+  /**
+   * @return {number}
+   */
+  _findNextNodeCompletionTime() {
+    let minimumTime = Infinity;
+    for (const node of this._nodesInProcess) {
+      minimumTime = Math.min(minimumTime, this._estimateTimeRemaining(node));
+    }
+
+    return minimumTime;
+  }
+
+  /**
+   * @param {!Node} node
+   * @param {number} timePeriodLength
+   * @param {number} totalElapsedTime
+   */
+  _updateProgressMadeInTimePeriod(node, timePeriodLength, totalElapsedTime) {
+    if (node.type !== Node.TYPES.NETWORK) throw new Error('Unsupported');
+
+    const timingData = this._nodeTiming.get(node);
+    const connection = this._connections.get(node.record.connectionId);
+    const calculation = connection.simulateDownloadUntil(
+      node.record.transferSize - timingData.bytesDownloaded,
+      timingData.timeElapsed,
+      timePeriodLength - timingData.timeElapsedOvershoot
+    );
+
+    connection.setCongestionWindow(calculation.congestionWindow);
+
+    if (timingData.estimatedTimeElapsed === timePeriodLength) {
+      timingData.endTime = totalElapsedTime;
+
+      connection.setWarmed(true);
+      this._connectionsInUse.delete(connection);
+
+      this._nodesCompleted.add(node);
+      this._nodesInProcess.delete(node);
+
+      for (const dependent of node.getDependents()) {
+        this._enqueueNodeIfPossible(dependent);
+      }
+    } else {
+      timingData.timeElapsed += calculation.timeElapsed;
+      timingData.timeElapsedOvershoot += calculation.timeElapsed - timePeriodLength;
+      timingData.bytesDownloaded += calculation.bytesDownloaded;
+    }
+  }
+
+  /**
+   * @return {number}
+   */
+  estimate() {
+    // initialize all the necessary data containers
+    this._initializeNetworkRecords();
+    this._initializeNetworkConnections();
+    this._initializeAuxiliaryData();
+
+    const nodesInQueue = this._nodesInQueue;
+    const nodesInProcess = this._nodesInProcess;
+
+    // add root node to queue
+    nodesInQueue.add(this._graph.getRootNode());
+
+    let depth = 0;
+    let totalElapsedTime = 0;
+    while (nodesInQueue.size || nodesInProcess.size) {
+      depth++;
+
+      // move all possible queued nodes to in process
+      for (const node of nodesInQueue) {
+        this._startNodeIfPossible(node, totalElapsedTime);
+      }
+
+      // set the available throughput for all connections based on # inflight
+      this._updateNetworkCapacity();
+
+      // find the time that the next node will finish
+      const minimumTime = this._findNextNodeCompletionTime();
+      totalElapsedTime += minimumTime;
+
+      // update how far each node will progress until that point
+      for (const node of nodesInProcess) {
+        this._updateProgressMadeInTimePeriod(
+          node,
+          minimumTime,
+          totalElapsedTime
+        );
+      }
+
+      if (depth > 10000) {
```

Comment:
While it was also very useful for early debugging with a small value, this is just a sensibility check to avoid looping forever and instead fail in the event of illogical network record or trace event relationships (or bug in graph construction though at this point it's synonymous with illogic event relationships given the interpretations we're taking).

Reply:
> this is just a sensibility check to avoid looping forever and instead fail in the event of illogical network record or trace event relationships

same as above, maybe we can move to an `assertValidDepGraph` method in the future